### PR TITLE
[RND-552] [RND-540] Introduce total ordering to document insert/updates - MongoDB and PostgreSQL

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Update.ts [hunk 1]
```diff
@@ -295,17 +295,17 @@ async function updateDocumentByDocumentUuidTransaction(
...
     meadowlarkId,
     edfiDoc,
-    validateDocumentReferencesExist,
-    security.clientId,
-    Date.now(),
-    lastModifiedAt,
-  );
+    validate: validateDocumentReferencesExist,
+    createdBy: security.clientId,
+    createdAt: null,
```
Comment (stephenfuqua) [inline Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Update.ts:306]: Restoring this to the actual created at: is that what you intended with the comment in an earlier file, about fixing the `createdAt` when replacing a document?

Comment (stephenfuqua) [reply Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Update.ts:306]: Yes, clearly that is the case. I got myself lost in the thread while bouncing between this review and something else.

File: Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Get.ts [hunk 2]
```diff
@@ -21,17 +21,18 @@ export async function getDocumentByDocumentUuid(
...
+    if (result === null) {
+      return { response: 'GET_FAILURE_NOT_EXISTS', edfiDoc: {}, documentUuid, lastModifiedDate: 0 };
+    }
+
     return {
       response: 'GET_SUCCESS',
-      document: { id: documentUuid, ...result.edfiDoc, _lastModifiedDate: documentLastModifiedDate },
+      edfiDoc: result.edfiDoc,
+      documentUuid,
+      lastModifiedDate: result.lastModifiedAt,
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Get.ts:32]: Changed GetResult in meadowlark-core to require id and lastModifiedDate explicitly.

File: Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts [hunk 3]
```diff
@@ -324,69 +327,91 @@ export async function insertOutboundReferences(
...
+        meadowlark_id = $1,
+        document_identity = $2,
+        project_name = $3,
+        resource_name = $4,
+        resource_version = $5,
+        is_descriptor = $6,
+        validated = $7,
+        created_by = $8,
+        edfi_doc = $9,
+        last_modified_at = $10
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts:357]: pg-format treats numeric parameters like strings, so switched to node-postgres built-in.

File: Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts [hunk 4]
```diff
@@ -324,69 +327,91 @@ export async function insertOutboundReferences(
...
+): Promise<boolean> {
+  const queryResult: QueryResult<any> = await client.query(
+    `
       INSERT INTO meadowlark.documents
         (meadowlark_id, document_uuid, document_identity, project_name, resource_name, resource_version, is_descriptor,
-        validated, created_by, edfi_doc)
-        VALUES (%L)
+        validated, created_by, edfi_doc, created_at, last_modified_at)
+        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)
         RETURNING document_uuid;`,
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts:398]: pg-format treats numeric parameters like strings, so switched to node-postgres built-in.

File: Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts [hunk 5]
```diff
@@ -324,69 +327,91 @@ export async function insertOutboundReferences(
...
-      documentValues[0],
-      documentValues[2],
-      documentValues[3],
-      documentValues[4],
-      documentValues[5],
-      documentValues[6],
-      documentValues[7],
-      documentValues[8],
-      documentValues[9],
-      documentValues[1],
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts:386]: This was not being explicit about values

File: Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts [hunk 6]
```diff
@@ -409,7 +434,7 @@ async function checkDeleteResult(
     return deleteResult;
   }
   if (countDeletedRows) {
-    return deleteQueryResult.rows[0].count === '0'
+    return deleteQueryResult.rows[0].count === 0
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts:437]: Better handling of numerics

File: Meadowlark-js/packages/meadowlark-core/src/handler/FrontendFacade.ts [hunk 7]
```diff
@@ -42,7 +43,8 @@ async function initialize(): Promise<void> {
 function postStack(): MiddlewareStack {
   return R.once(
     R.pipe(
-      authorize,
+      timestampRequest,
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/handler/FrontendFacade.ts:46]: Create request timestamp for 1st action

File: Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts [hunk 8]
```diff
@@ -324,69 +327,91 @@ export async function insertOutboundReferences(
 }
 
 /**
- * Inserts or updates a document in the database
- * @param param0 Document info for insert/update
- * @param isInsert is insert or update SQL re
- * @returns if the result returned rows
+ * Updates a document in the database
  */
-export async function insertOrUpdateDocument(
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/SqlHelper.ts:332]: Split function into two. Nothing in common worth keeping together.

**Pull request merged**